README.md

```diff
@@ -1,16 +1,1831 @@
+
+ send.result.ok: success
+ send.result.error: { error }
+ --
+
+ send.fx.fork: fork
+ send.fx.join: join
+}
+
+
+send.run --> send.task
+send.auth --> auth
+
+read.run --> read.task
+read.auth --> auth
+
+update.run --> update.task
+update.auth --> auth
+
+scope --> send.task
+scope --> read.task
+scope --> update.task
+
+
+send.promise.await --> send: ok
+send.promise.await --> send: error
+send.promise.await --> send: *
+
+read.promise.await --> read.task: ok
+read.promise.await --> read.task: error
+read.promise.await --> read.task: *
+
+
+send.receipt.out --> send.result
+send.receipt.sig --> send.receipt.ran
+send.receipt.sig --> send.receipt.out
+send.receipt.sig --> send.receipt.fx
+send.receipt.sig --> send.receipt.meta
+send.receipt.sig --> send.receipt.iss
+send.receipt.sig --> send.receipt.prf
+
+
+send.receipt.fx --> send.fx.fork
+send.receipt.fx --> send.fx.join
+send.fx.fork --> update.task
+send.fx.fork --> read.task
+read.cause --> send
+update.cause --> send
+send.receipt.ran --> send
+```
+
+### 2.2.1 Task
+
+A [Task] is like a deferred function application: a request to perform some action on a resource with specific input.
+
+### 2.2.2 Authorization
+
+An [Authorization] is a cryptographically signed proof permitting execution of referenced tasks. It allows invoker to authorize a group of tasks using one cryptographic signature.
+
+### 2.2.3 Invocation
+
+An [Invocation] is an invoker authorized instruction to the [Executor] to run the [Task].
+
+### 2.2.4 Result
+
+A [Result] is the output of a [Task].
+
+### 2.2.5 Receipt
+
+A [Receipt] is a cryptographically signed description of the [Invocation] output.
+
+### 2.2.6 Effect
+
+An [Effect] are the instruction to the [Executor] to run set of [Task]s concurrently.
+
+## 2.3 IPLD Schema
+
+```ipldsch
+type Task struct {
+  with    URI
+  do      Ability
+  input   {String: any}
+
+  nnc     string
+}
+
+type URI string
+type Ability string
+
+type Authorization struct {
+  # Authorization is denoted by the set of links been authorized
+  scope   [&Any]
+  # Scope signed by the invoker
+  s       VarSig
+}
+
+type Invocation struct {
+  v       SemVer
+
+  run     &Task
+  # Receipt of the invocation that caused this invocation
+  cause   optional &Invocation
+  # Task authorization.
+  auth    &Authorization
+
+  meta    {String : any}
+
+  prf     [&UCAN]
+}
+type SemVer string
+
+type Receipt struct {
+  # Invocation this is a receipt for
+  ran     &Invocation
+
+  # Output of the invocation
+  out     Result
+
+  # Effects to be performed
+  fx      Effect
+
+  # All the other metadata
+  meta    {String: any}
+
+  # Principal that issued this receipt. If omitted issuer is
+  # inferred from the invocation task audience.
+  iss     optional Principal
+
+  # When issuer is different from executor this MUST hold a UCAN
+  # delegation chain from executor to the issuer. Should be omitted when
+  # executor is an issuer.
+  prf     [&UCAN]
```

Comment:
I'm okay with including this, and I do recall that we've talked about this before, but it's slipping my mind so apologies. The concrete use case for this is when the executor subdelegates to another principal to actually run the job?

Reply:
Turns out in IPLD schema omitting list field from the model implies `[]` so `prf` is optional.

Use case is Executor delegates to the Worker to act on it's behalf, now if Worker signed  it needs to provide proof of delegation from `Executor`.

Both are optional and only relevant if `Receipt` is not signed by an Executor. I believe I go into more details on this under `Receipt` section of the spec.

Reply:
For consistency and tracing, wouldn't you want to keep that information around? In the other PR, there's a `rec` field that shows "here's the recursive invocation that I requested from another machine".

This accounts for the fact that delegation is not invocation in the trace. It also allows the system to operate across system boundaries when a machine gets a request that it cannot execute itself (e.g. it doens't haev the private key, doesn't have a GPU, etc) it can subdelegate that part of the job to another service, acting as an abstraction for a combined service. e.g. Fission wants to store data at web3storage, but update DNS on CloudFlare.

Reply:
That said, in the spirit of rough consensus, this isn't an "over my dead body" — just registering it as something to consider as we try this in production

Reply:
We work with a constraint that Executor key isn’t available at runtime so delegation can be made on demand. That is why it needs to be pre-arranged ahead of time and handed to the actor that will do the execution.

So it’s little bit like powerbox here too. Receipt signing is done by different key and to prove it had authority proof to prearrange delegation is provided.

Reply:
I don’t mind an additional field like `rec`, my version had an `origin` field for similar purposes , i just generalized it mean related receipt.

Regardless of the name we won’t be able to utilize it due to constraints of our system as we won’t be able to create receipts from Executor’s key.